internal/backend_gcs.go

```diff
@@ -0,0 +1,276 @@
+package internal
+
+import (
+	"cloud.google.com/go/storage"
+	"context"
+	"fmt"
+	. "github.com/kahing/goofys/api/common"
+	"google.golang.org/api/iterator"
+	"google.golang.org/api/option"
+	"strings"
+	"syscall"
+)
+
+type GCSBackend struct{
+	client *storage.Client
+	config *GCSConfig
+	cap *Capabilities
+	bucket *storage.BucketHandle
+	logger *LogHandle
+}
+
+
+// NewGCS initializes GCS client and returns GCSBackend
+func NewGCS(config *GCSConfig) (*GCSBackend, error){
+	var client *storage.Client
+	var err error
+
+	if config.Credentials != nil {
+		ctx := context.Background()
+		client, err = storage.NewClient(ctx)
+	} else {
+		ctx := context.Background()
+		client, err = storage.NewClient(ctx, option.WithoutAuthentication())
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &GCSBackend{
+		client: client,
+		config: config,
+		bucket: client.Bucket(config.Bucket),
+		cap: &Capabilities{
+			MaxMultipartSize: 5 * 1024 * 1024 * 1024,
+			Name:             "gcs",
+			// TODO: no parallel multipart but resumable uploads
+			NoParallelMultipart: false,
+		},
+		logger: GetLogger("gcs"),
+	}, nil
+}
+
+func (g *GCSBackend) Init(key string) (err error) {
+	g.logger.Debug("Initializes GCS")
+	err = g.testBucket(key)
+
+	return
+}
+
+func (g *GCSBackend) testBucket(key string) (err error) {
+	ctx := context.Background()
+
+	// v01: Require users to have read access to the bucket bucket.get)
+	g.logger.Debug("Get Bucket Info")
+	_, err = g.bucket.Attrs(ctx)
+	if err != nil {
+		return err
+	}
+
+	g.logger.Debug("Get Object Info")
+	_, err = g.HeadBlob(&HeadBlobInput{Key: key})
+	if err == storage.ErrObjectNotExist {
+		err = nil
+	}
+
+	return
+}
+
+func (g *GCSBackend) Capabilities() *Capabilities {
+	return g.cap
+}
+
+// Bucket() typically returns the bucket/prefix
+func (g *GCSBackend) Bucket() string {
+	return g.config.Bucket
+}
+
+func getResponseStatus(err error) string {
+	if err != nil {
+		return fmt.Sprintf("ERROR: %v", err)
+	}
+	return "SUCCESS"
+}
+
+func (g *GCSBackend) HeadBlob(param *HeadBlobInput) (*HeadBlobOutput, error) {
+	attrs, err := g.bucket.Object(param.Key).Attrs(context.Background())
+	g.logger.Debugf("HEAD %v = %v", param.Key, getResponseStatus(err))
+
+	if err != nil {
+		return nil, err
+	}
+
+	var metadata map[string]*string
+	if attrs.Metadata != nil {
+		metadata = mapGCSMetadataToBackendMetadata(attrs.Metadata)
+	}
+
+	return &HeadBlobOutput{
+		BlobItemOutput: BlobItemOutput{
+			Key: &attrs.Name,
+			ETag: &attrs.Etag,
+			LastModified: &attrs.Updated,
+			Size: uint64(attrs.Size),
+			StorageClass: &attrs.StorageClass,
+		},
+		ContentType: &attrs.ContentType,
+		IsDirBlob: strings.HasSuffix(param.Key, "/"),
+		Metadata: metadata,
+	}, nil
+}
+
+func mapGCSMetadataToBackendMetadata(m map[string]string) map[string]*string {
+	newMap := make(map[string]*string)
+
+	for k, v := range m {
```

Comment:
why toLower?

Reply:
i think i just followed https://github.com/kahing/goofys/blob/45b8d78375af1b24604439d2e60c567654bcdf88/internal/backend_s3.go#L331 where they tried to normalize the metadata keys into having lowercase keys